dali/kernels/audio/mel_scale/mel_filter_bank_gpu.cu

```diff
@@ -0,0 +1,311 @@
+    for (uint64_t ti = 0; ti < nframes_.size(); ++ti) {
+      const T *in = in_list[ti];
+      T *out = out_list[ti];
+      for (int64_t s = 0; s < nframes_[ti]; ++s) {
+        auto nblocks = (nwindows_[ti] + kBlockDim2 - 1) / kBlockDim2;
+        for (int b = 0; b < nblocks; ++b) {
+          block_descs_[block_id].in_frame = in;
+          block_descs_[block_id].out_frame = out;
+          ++block_id;
+        }
+        in += nwindows_[ti] * fft_dim_;
+        out += nwindows_[ti] * args_.nfilter;
+      }
+    }
+  }
+
+  void FillBlockDescsInnerFft(const T* const* in_list, T **out_list) {
+    int block_id = 0;
+    int sample = 0;
+    while (block_id < block_descs_.size()) {
+      auto sample_size = block_descs_[block_id].out_frame_size;
+      auto nblocks = (sample_size + kBlockDim1 - 1) / kBlockDim1;
+      for (int i = 0; i < nblocks; ++i, ++block_id) {
+        block_descs_[block_id].in_frame = in_list[sample];
+        block_descs_[block_id].out_frame = out_list[sample];
+      }
+      ++sample;
+    }
+  }
+
+  std::vector<int64_t> interval_ends_;
+  std::vector<int64_t> nframes_;
+  std::vector<int64_t> nwindows_;
+  std::vector<BlockDesc<T>> block_descs_;
+  int64_t fft_dim_;
+  USE_MEL_FILTER_IMPL_MEMBERS(T, Dims);
+};
+
+template <typename T, int Dims>
+KernelRequirements MelFilterBankGpu<T, Dims>::Setup(KernelContext &context,
+                                                    const InListGPU<T, Dims> &in,
+                                                    const MelFilterBankArgs &original_args) {
+  auto args = original_args;
+  args.axis = args.axis >= 0 ? args.axis : Dims - 2;
+  TensorListShape<Dims> out_shape(in.shape.num_samples());
+  for (int64_t s = 0; s < out_shape.num_samples(); ++s) {
+    auto s_shape = in.shape[s];
+    s_shape[args.axis] = args.nfilter;
+    out_shape.set_tensor_shape(s, s_shape);
+  }
+  KernelRequirements req;
+  req.output_shapes = {out_shape};
+
+  auto fftdim = in.shape[0][args.axis];
+  for (int s = 1; s < in.shape.nsamples; ++s) {
+    DALI_ENFORCE(in.shape[s][args.axis] == fftdim,
+        "All samples should have the same FFT dimension");
+  }
+  ScratchpadEstimator se;
+  args.nfft = args.nfft > 0 ? args.nfft : 2 * (in.shape[0][args.axis] - 1);
+  args.freq_high = args.freq_high > 0 ? args.freq_high : args.sample_rate / 2;
+  if (!impl_ || impl_->Args() != args) {
+    impl_.reset();
+    switch (args.mel_formula) {
+      case MelScaleFormula::HTK:
+        impl_ = std::make_unique<Impl>(HtkMelScale<T>(), args, se, in.shape);
+        break;
+      case MelScaleFormula::Slaney:
+      default:
+        impl_ = std::make_unique<Impl>(SlaneyMelScale<T>(), args, se, in.shape);
+        break;
+    }
+  }
+  req.scratch_sizes = se.sizes;
+  return req;
+}
+
+template <typename T, int Dims>
+void MelFilterBankGpu<T, Dims>::Run(
+    KernelContext &context,
+    OutListGPU<T, Dims> &out,
+    const InListGPU<T, Dims> &in,
+    const MelFilterBankArgs &original_args) {
```

Comment:
I think we can just stop requiring Run and Setup to have the same API - we don't have any automation where this would help, and with handwritten code it does more ill than good. So... this argument can be safely dropped.

Reply:
done